### PR TITLE
Improve `compiler/utils/bitsets.nim` module

### PR DESCRIPTION
**Pull request by zerbina**
* Add tests for all functions (there were none previously)
* Turn procs into funcs
* All `bitset` functions use set views now. This makes them
  usable in cases where the bit-set is not stored as a `seq`
* Add `bitSetInclRange`, which aims to be more efficient than a
  simple `for x in a..b: s.bitSetIncl(x)`
* Add tests for `bitSetInclRange`
* Adjust functions in `nimsets` to take bit-set views as arguments

The `bitSetContains(x, y)` function previously calculated if "x is a
subset of y" instead of if "y is a subset of x". This is now changed
in order to make `bitSetContains` consistent with `contains`

### CODE REVIEW

File: compiler/utils/bitsets.nim [hunk 1]
```diff
@@ -15,58 +15,108 @@ type
   TBitSet* = seq[ElemType]    # we use byte here to avoid issues with
                               # cross-compiling; uint would be more efficient
                               # however
+  TBitSetView* = openArray[ElemType] ## A view into a bit set. This doesn't
+                                     ## necessarily have to be a `TBitSet`,
+                                     ## but can be any byte sequence storing
+                                     ## representing a bit set
```
Comment (alaviss) [inline compiler/utils/bitsets.nim:21]: If you want to alias `openArray[X]`, you have to write it out explicitly for now as the compiler might interpret this as a concrete view type, which is an unstable feature.

Comment (zerbina) [reply compiler/utils/bitsets.nim:21]: Even when only used as a parameter type? That's a bit annoying, but I'll use `openArray[ElemType]` for now then

File: compiler/ast/nimsets.nim [hunk 2]
```diff
@@ -64,10 +64,11 @@ proc someInSet*(s: PNode, a, b: PNode): bool =
         return true
   result = false
 
-proc toBitSet*(conf: ConfigRef; s: PNode): TBitSet =
+proc inclTreeSet*(result: var openArray[ElemType], conf: ConfigRef; s: PNode) =
```
Comment (alaviss) [inline compiler/ast/nimsets.nim]: You're probably missing an import or so?

**Review by alaviss — COMMENTED** [review]
LGTM, can merge once the compilation failure is fixed.

File: compiler/utils/bitsets.nim [hunk 3]
```diff
@@ -11,62 +11,116 @@
...
+
+func bitSetInit*(b: var TBitSet, length: int) =
+  ## Creates a new bitset in `b` with a size-in-bytes of `length * ElemSize` .
+  ## The resulting bitset allows for elements in the
+  ## range `0..<(length * ElemSize * 8)`
   newSeq(b, length)
 
-proc bitSetUnion*(x: var TBitSet, y: TBitSet) =
+func bitSetUnion*(x: var openArray[ElemType], y: openArray[ElemType]) =
+  ## Calculate the union between sets `x` and `y
```
Comment (alaviss) [inline compiler/utils/bitsets.nim]:
```suggestion
  ## Calculate the union between sets `x` and `y`
```

That should fix docgen...

Comment (zerbina) [reply compiler/utils/bitsets.nim]: Ah, thanks. I wasn't able to figure out why docgen was failing while all other test passed.

Should I revert the `TBitSetView` -> `openArray[ElemType]` change then, as it didn't cause the issue?

Comment (alaviss) [reply compiler/utils/bitsets.nim]: Feel free to

**Review by saem — APPROVED** [review]
Bors r+

Woot, this is gaining steam!
